opm/input/eclipse/Schedule/Action/ActionParser.cpp

```diff
@@ -17,324 +17,583 @@
+        this->advanceTokenPosition(); // Consume "("
 
-    if (curr.type == TokenType::open_paren) {
-        this->next();
-        auto inner_expr = this->parse_or();
+        const auto inner_expr = this->parse_or();
 
-        curr = this->current();
-        if (curr.type != TokenType::close_paren) {
-            return ASTNode { TokenType::error };
+        if (this->currentType() != Opm::Action::TokenType::close_paren) {
+            return Opm::Action::ASTNode { Opm::Action::TokenType::error };
         }
 
-        this->next();
+        this->advanceTokenPosition(); // Consume ")"
 
-        return ASTNode { inner_expr };
+        return inner_expr;
     }
-    else {
-        auto left_node = this->parse_left();
-        if (left_node.type == TokenType::error) {
-            return ASTNode { TokenType::error };
-        }
 
-        auto op_node = this->parse_op();
-        if (op_node.type == TokenType::error) {
-            return ASTNode { TokenType::error };
-        }
-
-        auto right_node = this->parse_right();
-        if (right_node.type == TokenType::error) {
-            return ASTNode { TokenType::error };
-        }
+    auto left_node = this->parse_left();
+    if (left_node.type == Opm::Action::TokenType::error) {
+        return Opm::Action::ASTNode { Opm::Action::TokenType::error };
+    }
 
-        op_node.add_child(std::move(left_node));
-        op_node.add_child(std::move(right_node));
+    auto op_node = this->parse_op();
+    if (op_node.type == Opm::Action::TokenType::error) {
+        return Opm::Action::ASTNode { Opm::Action::TokenType::error };
+    }
 
-        return op_node;
+    auto right_node = this->parse_right();
+    if (right_node.type == Opm::Action::TokenType::error) {
+        return Opm::Action::ASTNode { Opm::Action::TokenType::error };
     }
+
+    op_node.add_child(std::move(left_node));
+    op_node.add_child(std::move(right_node));
+
+    return op_node;
 }
 
-ASTNode Parser::parse_and()
+Opm::Action::ASTNode ActionParser::parse_and()
 {
     auto left = this->parse_cmp();
-    if (left.type == TokenType::error) {
-        return ASTNode { TokenType::error };
+    if (left.type == Opm::Action::TokenType::error) {
+        return Opm::Action::ASTNode { Opm::Action::TokenType::error };
     }
 
-    auto curr = this->current();
-    if (curr.type == TokenType::op_and) {
-        ASTNode and_node(TokenType::op_and);
-        and_node.add_child(std::move(left));
-
-        while (this->current().type == TokenType::op_and) {
-            this->next();
-            auto next_cmp = this->parse_cmp();
-            if (next_cmp.type == TokenType::error) {
-                return ASTNode { TokenType::error };
-            }
-
-            and_node.add_child(std::move(next_cmp));
+    if (this->currentType() != Opm::Action::TokenType::op_and) {
+        return left;
+    }
+
+    auto and_node = Opm::Action::ASTNode { Opm::Action::TokenType::op_and };
```

Comment:
return above and this block could potentially be combined to enable nrvo. not sure if it matters though and it would probably impact readability.

Reply:
> return above and this block could potentially be combined to enable nrvo.

Possibly.  It's not immediately clear to me how, though, because the `left` node in the return statement above is something other than `op_and` by construction.

Reply:
afaik gcc will enable nrvo if all return statements refer to the same stack object. but anyways, was just a drive-by.

Reply:
> afaik gcc will enable nrvo if all return statements refer to the same stack object. but anyways, was just a drive-by.

Okay.  We'll keep this in mind for future refinements.